File: coteacher-server/usecase/teacher/teacher.go
package teacher

import (
	"connectrpc.com/connect"
	"golang.org/x/exp/slog"

	"context"

	pb "github.com/KinjiKawaguchi/Coteacher/proto-gen/go/coteacher/v1"

	"github.com/KinjiKawaguchi/Coteacher/coteacher-server/domain/repository/ent"
	entteacher "github.com/KinjiKawaguchi/Coteacher/coteacher-server/domain/repository/ent/teacher"
	entuser "github.com/KinjiKawaguchi/Coteacher/coteacher-server/domain/repository/ent/user"

	"github.com/google/uuid"
)

type Interactor struct {
	entClient *ent.Client
	logger    *slog.Logger
}

func NewInteractor(entClient *ent.Client, logger *slog.Logger) *Interactor {
	return &Interactor{entClient, logger}
}

func (i *Interactor) CheckTeacherExistsByID(ctx context.Context, req *pb.CheckTeacherExistsByIDRequest) (*pb.CheckTeacherExistsByIDResponse, error) {
	teacherID, err := uuid.Parse(req.Id)
	if err != nil {
		return nil, connect.NewError(connect.CodeInvalidArgument, err)
	}

	exists, err := i.entClient.Teacher.Query().
		Where(entteacher.ID(teacherID)).
		Exist(ctx)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	return &pb.CheckTeacherExistsByIDResponse{
		Exists: exists,
	}, nil
}

func (i *Interactor) CheckTeacherExistsByEmail(ctx context.Context, req *pb.CheckTeacherExistsByEmailRequest) (*pb.CheckTeacherExistsByEmailResponse, error) {
	//EmailがUserTableでマッチするUserIDががTeacherIDに存在するかを返す

	id := i.entClient.User.Query().
		Where(entuser.Email(req.Email)).
		Select(entuser.FieldID).
		FirstIDX(ctx)

	exists, err := i.entClient.Teacher.Query().
		Where(entteacher.ID(id)).
		Exist(ctx)
	if err != nil {
		return nil, connect.NewError(connect.CodeInternal, err)
	}

	return &pb.CheckTeacherExistsByEmailResponse{
		Exists: exists,
	}, nil
}

File: coteacher-app/src/repository/teacher.ts
import { backendGrpcTransport } from '@/config/connectRpc';
import { TeacherService } from '@/gen/proto/coteacher/v1/teacher_connect';
import {
  CheckTeacherExistsByEmailRequest,
  CheckTeacherExistsByIDRequest,
} from '@/gen/proto/coteacher/v1/teacher_pb';
import {
  PromiseClient,
  Transport,
  createPromiseClient,
} from '@bufbuild/connect';

class TeacherRepository {
  readonly cli: PromiseClient<typeof TeacherService>;

  constructor(t: Transport) {
    this.cli = createPromiseClient(TeacherService, t);
  }

  async checkTeacherExistsByID(teacherId: string | null): Promise<boolean> {
    const req = new CheckTeacherExistsByIDRequest();
    if (teacherId) {
      req.id = teacherId;
    } else {
      req.id = localStorage.getItem('UserID') || '';
    }
    const res = await this.cli.checkTeacherExistsByID(req);
    return res.exists;
  }

  async checkTeacherExistsByEmail(email: string | null): Promise<boolean> {
    const req = new CheckTeacherExistsByEmailRequest();
    if (email) {
      req.email = email;
    } else {
      req.email = localStorage.getItem('UserEmail') || '';
    }
    const res = await this.cli.checkTeacherExistsByEmail(req);
    return res.exists;
  }
}

export const teacherRepo = new TeacherRepository(backendGrpcTransport);
